compiler/rustc_trait_selection/src/solve/trait_goals/structural_traits.rs

```diff
@@ -235,3 +241,114 @@ pub(crate) fn extract_tupled_inputs_and_output_from_callable<'tcx>(
         }
     }
 }
+
+/// Assemble a list of predicates that would be present on a theoretical
+/// user impl for an object type. These predicates must be checked any time
+/// we assemble a built-in object candidate for an object type, since they
+/// are not implied by the well-formedness of the type.
+///
+/// For example, given the following traits:
+///
+/// ```rust,ignore (theoretical code)
+/// trait Foo: Baz {
+///     type Bar: Copy;
+/// }
+///
+/// trait Baz {}
+/// ```
+///
+/// For the dyn type `dyn Foo<Item = Ty>`, we can imagine there being a
+/// pair of theoretical impls:
+///
+/// ```rust,ignore (theoretical code)
+/// impl Foo for dyn Foo<Item = Ty>
+/// where
+///     Self: Baz,
+///     <Self as Foo>::Bar: Copy,
+/// {
+///     type Bar = Ty;
+/// }
+///
+/// impl Bar for dyn Foo<Item = Ty> {}
+/// ```
+///
+/// However, in order to make such impls well-formed, we need to do an
```

Comment:
That's the main concerning step here :thinking: because I don't have a strong understanding of **why** the impl with unnormalized projections is not well-formed.

It doesn't compile with either the new solver or the old one, but that doesn't feel like a good enough explanation :< can look into this a bit myself (and should have a rustc-dev-guide section about this)

Reply:
```rust
trait Trait {
    type Assoc: Bound;
}

trait Bound {}

struct Ty;

/*

impl Trait for dyn Trait<Assoc = Ty> {
    type Assoc = Ty;
}

*/

struct DynTrait;

impl Trait for DynTrait
where
    <DynTrait as Trait>::Assoc: Bound,
{
    type Assoc = Ty;
}

fn needs_trait(_: impl Trait) {}

fn main() {
    needs_trait(DynTrait);
}
```

So this actually isn't (technically) failing during wfcheck. On the "classic" solver, this overflows in `normalize_param_env_or_error` in the `param_env` query for the impl. :thinking:

Reply:
lcnr/solver-woes#10 to track this